source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -482,11 +482,193 @@ account:
+
+    ...
+
+The ``concordium_quickcheck`` macro takes the ``num_tests`` attribute for specifying the number of random tests to run.
+In the code snippet below, the parameters ``address`` and ``amount`` are generated randomly.
+The process of generating random input and running the test is repeated ``num_tests = 500``.
+
+.. code-block:: rust
+
+    #[concordium_cfg_test]
+    mod test {
+
+       #[concordium_quickcheck(num_tests = 500)]
+       fn some_property_test(address: Address, amount: Amount) -> bool {
+        ...
+        // Instantiate custom struct with random parameters, if necessary.
+        let input = MyParameters { sender: address, payment: amount }
+        ...
+        }
+    }
+
+The types ``Address`` and ``Amount`` in the example have ``Arbitrary`` trait implementations, which are used to obtain random values.
+Read more about available ``Arbitrary`` instances for Concordium-specific types in |concordium_contracts_common|_ documentation.
+|QuickCheck|_ defines ``Arbitrary`` instances for standard data types, like numbers and collections (``Vec``, ``BTreeMap``, etc.).
+These instances are available by default when writing tests.
+Custom user data type instances, like ``MyParameters`` above, can be created directly in tests using the random input parameters or by defining ``Arbitrary`` instances.
+See more details on QuickCheck's ``Arbitrary`` `here <https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html>`_.
+
+The same command is used for running Wasm QuickCheck tests as in :ref:`tests_in_wasm`:
+
+.. code-block:: console
+
+    $cargo concordium test
+
+When a test fails, it reports the random seed used to produce the input values.
+After making the required fixes to the code, you can use the same seed to see whether the previously failed tests work on the same generated values.
+The seed is a ``u64`` number, which can be provided along with the test command:
+
+.. code-block:: console
+
+    $cargo concordium test --seed 1234567890
+
+Concordium QuickCheck tests can also be run with:
+
+.. code-block:: console
+
+    $cargo test
+
+By default, this command compiles the contract, unit tests, and QuickCheck tests to machine code for your local target (most likely x86_64) and runs them.
+
+.. note::
+
+    Printing and supplying a seed is only possible using ``cargo concordium test``.
+
+.. warning::
+
+    Avoid using ``fail!`` and ``claim!`` variants in ``#[concordium_quickcheck]`` tests.
+    In Wasm unit tests (see :ref:`tests_in_wasm`) these commands report an error.
+    However, using them in QuickCheck tests makes the tests fail without providing a counterexample when running with ``cargo concordium test``.
+    Also avoid using ``assert_eq!``, ``panic!`` or any other command that panics.
+    Return a boolean value instead.
+
+Example
+-------
+
+Consider a counter with a threshold: if the count is less than the threshold, it gets incremented; otherwise, it stays unchanged.
```

Comment:
While this example is good warm-up example, I would prefer if we have a more realistic one.

Is there something in the piggy-bank example that we could specify like this?

Reply:
I looked at `piggy-bank` with exactly this purpose. But all I could come up with looks as artificial as this example. However, once we integration testing, we could make a nice story about `piggy-bank` being a transition system and show some temporal properties, e.g. reachability of states over an arbitrary sequence of calls.

wCCD could be a possibility, but these examples would be more complex and require more background. I wanted to make a separate task of adding property tests to existing examples and then writing up some sort of tutorial, maybe explaining the methodology in more detail and showing how to write custom `Arbitrary`. Does it make sense?

Reply:
While the current example is contrived I think it shows enough info to get started **if** you know something about property-based testing already. Perhaps we should link to some general information about property-based testing. Do you know any good beginner resources?

And I think a separate tutorial is a good idea.

Reply:
I don't think I have enough time to pick (or evaluate) good resources on property-based testing, if we plan to finish up with the PR today-ish. I can add *some* links, of course, just as initial pointers.

Reply:
Ok, the link was just a suggestion, we can do it later 😊

Reply:
Is it ready for another review?

Reply:
@Bargsteen yep, requested re-review :)